src/header-parser.ts

```diff
@@ -0,0 +1,336 @@
+import type Spec from './Spec';
+
+export function parseStructuredHeaderH1(
+  spec: Spec,
+  header: Element
+): { name: string | null; formattedHeader: string | null; formattedParams: string | null } {
+  // parsing is intentionally permissive; the linter can do stricter checks
+  // TODO have the linter do checks
+  let headerText = header.innerHTML;
+  const prefix = headerText.match(/^\s*(Static|Runtime) Semantics:\s*/);
+  if (prefix != null) {
+    headerText = headerText.substring(prefix[0].length);
+  }
+
+  let formattedHeader = null;
+  let formattedParams = null;
+
+  const parsed = headerText.match(/^\s*([^(\s]+)\s*\((.*)\)\s*$/s);
+  if (parsed == null) {
+    spec.warn({
+      type: 'contents',
+      ruleId: 'header-format',
+      message: `failed to parse header`,
+      node: header,
+      nodeRelativeColumn: 1,
+      nodeRelativeLine: 1,
+    });
+    return { name: null, formattedHeader, formattedParams };
+  }
+
+  type Param = { name: string; type: string | null };
+  const name = parsed[1];
+  let paramText = parsed[2].trim();
+  const params: Array<Param> = [];
+  const optionalParams: Array<Param> = [];
+
+  if (/\(\s*\n/.test(headerText)) {
+    // if it's multiline, parse it for types
+    const paramChunks = paramText
+      .split('\n')
+      .map(c => c.trim())
+      .filter(c => c.length > 1);
+    let index = 0;
+    for (const chunk of paramChunks) {
+      ++index;
+      // TODO linter enforces all optional params come after non-optional params
+      const parsedChunk = chunk.match(/^(optional\s*)?([A-Za-z0-9_]+)\s*:\s*(\S.*\S)/);
+      if (parsedChunk == null) {
+        spec.warn({
+          type: 'contents',
+          ruleId: 'header-format',
+          message: `failed to parse header (parameter ${index})`,
+          node: header,
+          nodeRelativeColumn: 1,
+          nodeRelativeLine: 1,
+        });
+        continue;
+      }
+      const optional = parsedChunk[1] != null;
+      const paramName = parsedChunk[2];
+      let paramType = parsedChunk[3];
+      if (paramType.endsWith(',')) {
+        paramType = paramType.slice(0, -1);
+      }
+      (optional ? optionalParams : params).push({
+        name: paramName,
+        type: paramType === 'unknown' ? null : paramType,
+      });
+    }
+    const formattedprefix = prefix == null ? '' : prefix[0].trim() + ' ';
+    formattedHeader = `${formattedprefix}${name} (${params.map(n => ' ' + n.name).join(',')}`;
+    if (optionalParams.length > 0) {
+      formattedHeader +=
+        optionalParams
+          .map((p, i) => ' [ ' + (i > 0 || params.length > 0 ? ', ' : '') + p.name)
+          .join('') + optionalParams.map(() => ' ]').join('');
+    }
+    formattedHeader += ' )';
+  } else {
+    let optional = false;
+    while (true) {
+      if (paramText.length == 0) {
+        break;
+      }
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      let { success, text, match } = eat(paramText, /^\s*\[(\s*,)?/);
```

Comment:
A more elegant solution might just surround each of these `eat` calls (and the successive statements) in blocks so you don't have to worry about declaring an unused var here.

Reply:
Ehh, I'm not going to worry about it. I kind of like all calls to `eat` having the same form.